WonderMake/Utilities/Container/ContainerBackend.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+#include "Utilities/Container/Traits/ContainerTraits.h"
+#include "Utilities/TypeTraits/ContainsType.h"
+#include <type_traits>
+
+template<typename TStorage>
+class ContainerBackend
+{
+public:
+	using StorageBackend = typename ContainerBackend<TStorage>;
+	using IteratorType = typename TStorage::iterator;
+	using ConstIteratorType = typename TStorage::const_iterator;
+
+	void Clear()
+	{
+		myBackend.clear();
+	}
+
+	size_t Count() const
+	{
+		return myBackend.size();
+	}
+	
+	TStorage myBackend;
+};
+
+
+template<typename TContainerBackend, typename TTrait>
+class ImplementTrait {};
+
+template<typename TContainerBackend>
+class ImplementTrait<TContainerBackend, Iterable>
+{
+	using IteratorType = typename TContainerBackend::iterator;
+	using ConstIteratorType = typename TContainerBackend::const_iterator;
+	using ElementType = typename TContainerBackend::value_type;
+
+	virtual typename IteratorType Erase(IteratorType aIt) = 0;
+	virtual typename IteratorType Erase(ConstIteratorType aIt) = 0;
+
+	virtual size_t Erase(const ElementType& aElementType) = 0;
+
+	virtual typename ConstIteratorType cbegin() const = 0;
+	virtual typename ConstIteratorType cend() const = 0;
+
+	virtual typename IteratorType begin() = 0;
+	virtual typename IteratorType end() = 0;
+};
+
+template<typename TContainerBackend>
+class ImplementTrait<TContainerBackend, Indexable>
+{
+	using IteratorType = typename TContainerBackend::iterator;
+	using ConstIteratorType = typename TContainerBackend::const_iterator;
+	using ElementType = typename TContainerBackend::value_type;
+	virtual ElementType& operator[](const size_t aIndex) = 0;
+	virtual const ElementType& operator[](const size_t aIndex) const = 0;
+
+	virtual IteratorType EraseAt(const size_t aIndex) = 0;
+};
+
+template<typename TContainerBackend, typename ... TTraits>
+class ImplementTraits : ImplementTrait<TContainerBackend, TTraits>...
+{
+public:
+	template<typename ... TOtherTraits>
+	static constexpr bool SatisfiesTraits()
+	{
+		return IsSubsetOf<ParameterPack<TOtherTraits...>, ResolvedImplications<TTraits...>::type>::value;
+	}
+
+	template<typename TOtherTraits> 
+	static constexpr bool SatisfiesTraitsPack()
+	{
+		return IsSubsetOf<TOtherTraits, ResolvedImplications<TTraits...>::type>::value;
+	}
+};
+
+
+template <typename TTraits, typename...>
+struct FirstSatisfyingBackend
+{
+	using type = void;
+	static constexpr bool found = false;
```

Comment:
Unless I'm mistaken, our code standard says these values should be capitalized. However, I guess this was done to mimic std?

Reply:
Ya my b